src/main/java/io/cdap/plugin/spark/HierarchyFlattener.java

```diff
@@ -147,69 +148,70 @@ public HierarchyFlattener(HierarchyConfig config) {
        The approach is to take N passes through the hierarchy, where N is the maximum depth of the tree.
        In iteration i, paths of length i - 1 are found.
 
-       The first step is to generate the 0 distance paths.
+       The first step is to generate paths with length 1.
 
        With input rows:
 
-         [parent:1, child:1, category:grocery]
          [parent:1, child:2, category:vegetable]
          [parent:1, child:3, category:dairy]
          [parent:2, child:5, category:lettuce]
          [parent:2, child:6, category:tomato]
-         [parent:4, child:4, category:fruit]
          [parent:4, child:6, category:tomato]
+         [parent:7, child:8, category:water]
 
        the starting points are:
 
-         [parent:1, child:1, level:0, root:true, leaf:0, category:grocery]
-         [parent:2, child:2, level:0, root:false, leaf:0, category:vegetable]
-         [parent:3, child:3, level:0, root:false, leaf:0, category:dairy]
-         [parent:4, child:4, level:0, root:true, leaf:0, category:fruit]
-         [parent:5, child:5, level:0, root:false, leaf:0, category:lettuce]
-         [parent:6, child:6, level:0, root:false, leaf:0, category:tomato]
+         [parent:1, child:2, level: 1, leaf:0, category:vegetable]
+         [parent:1, child:3, level: 1, leaf:0, category:dairy]
+         [parent:2, child:5, level: 1, leaf:0, category:lettuce]
+         [parent:2, child:6, level: 1, leaf:0, category:tomato]
+         [parent:4, child:6, level: 1, leaf:0, category:tomato]
+         [parent:7, child:8, level: 1, leaf:0, category:water]
+
      */
-    LOG.info("Starting computation for level 0");
-    Dataset<Row> currentLevel = getStartingPoints(input, dataFieldNames);
+    LOG.info("Starting computation for level 1");
+    Dataset<Row> currentLevel = getStartingPoints(input, dataFieldNames, startsWith);
+      // Remove self referencing nodes to avoid loops.
+    currentLevel = currentLevel.where(input.col(parentCol).notEqual(input.col(childCol)));
+
+    // Add path and connect by root columns to the dataset
```

Comment:
would be good to enhance the example in the comments above and add comments here to show what the path and connectByRoot columns would do

Reply:
done